pagination

snopoke opened the pull request:
@millerdev 

cc @dimagi/scale-team 

This PR has gotten pretty big. Happy to split out the SQL and reindex stuff out into a separate PR if necessary.
- refactor paginate_view into generic pagination function
- refactor resumable couch iterators and doc processors to be DB agnostic
- create iterators for SQL models
- convert SQL reindexers to new style bulk versions

probably easiest review per commit :fish:

millerdev commented on corehq/util/couch_helpers.py:
```diff
@@ -165,24 +129,21 @@ def paginate_view(db, view_name, chunk_size,
+        view_kwargs['startkey'] = result['key']
+        view_kwargs['startkey_docid'] = result['id']
+        view_kwargs['skip'] = 1
+        return [db, view_name], view_kwargs
```
Minor: could simplify this a bit.
```py
def next_args(result):
    if result:
        view_kwargs['startkey'] = result['key']
        view_kwargs['startkey_docid'] = result['id']
        view_kwargs['skip'] = 1
    return [db, view_name], view_kwargs
```

millerdev commented on corehq/util/pagination.py:
```diff
@@ -0,0 +1,55 @@
+        self.handlers = handlers or []
+
+    def add_handler(self, handler):
+        self.handlers.add(handler)
```
should probably be `self.handlers.append(handler)`

millerdev commented on corehq/util/pagination.py:
```diff
@@ -0,0 +1,55 @@
+
+    def page_start(self, total_emitted, *args, **kwargs):
+        for handler in self.handlers:
+            handler.page_start(self, total_emitted, *args, **kwargs)
```
I'm not understanding why this is passing `self`. Is that a bug? (same below)

snopoke replied:
yea, that's a bug

millerdev commented:
👍 looking good so far

snopoke committed:
refactor paginate view into generic pagination function

snopoke committed:
resumable function iterator

snopoke committed:
create ArgsProvider class to make contract clear

snopoke committed:
add resume support

snopoke committed:
refactor ResumableDocsByTypeIterator to use ResumableFunctionIterator

snopoke committed:
fix tests

The need for this update is that previously we were only storing
'startkey' and 'endkey' in the resume state. Now we also store
'startkey_docid' and 'skip'.

snopoke committed:
fix rebase

dimagimon commented on corehq/util/couch_doc_processor.py:
```diff
@@ -1,17 +1,39 @@
+
+    def get_next_args(self, result, *last_args, **last_view_kwargs):
+        try:
+            return super(ResumableDocsByTypeArgsProvider, self).get_next_args(result, *last_args, **last_view_kwargs)
```
E501 line too long (117 > 115 characters)

snopoke committed:
don't pass couchdb with every doc

snopoke committed:
extract progress logging to class

snopoke committed:
extract DocumentProvider class

snopoke committed:
refactor into DB agnostic document processors

snopoke committed:
fix test

snopoke committed:
rename module

snopoke committed:
pep8

snopoke committed:
rename test module

snopoke committed:
remove unique_key from doc processor

snopoke committed:
fix reindexer

snopoke committed:
make document provider abstract

snopoke committed:
pass list of doc types directly to doc provider and logger

snopoke committed:
make reindexer DB agnostic

snopoke committed:
fix doc_types in ReportXFormReindexer

snopoke committed:
split module up

snopoke committed:
sql doc processor classes

snopoke committed:
tests for resumable sql iterators

snopoke committed:
update sql form reindexer

snopoke committed:
fix logging

snopoke committed:
convert sql case reindexer

snopoke committed:
convert ledger v2 reindexer

dimagimon commented on corehq/util/doc_processor/couch.py:
```diff
@@ -0,0 +1,128 @@
+
+    def get_next_args(self, result, *last_args, **last_view_kwargs):
+        try:
+            return super(ResumableDocsByTypeArgsProvider, self).get_next_args(result, *last_args, **last_view_kwargs)
```
E501 line too long (117 > 115 characters)

snopoke committed:
lint

snopoke commented on corehq/util/pagination.py:
```diff
@@ -71,8 +84,11 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
-            args, kwargs = next_args_fn(item, *args, **kwargs)
+        item = item if len_results else None
+        try:
+            args, kwargs = args_provider.get_next_args(item, *args, **kwargs)
```
also changed so that iteration continues until `get_next_args` raises `StopIteration`. The purpose of this is to support use cases like `DocTypeIterator` which iterates over multiple doc types in succession (basically a nested loop).

snopoke commented on corehq/util/doc_processor/sql.py:
```diff
@@ -0,0 +1,90 @@
+        )
+
+    def get_total_document_count(self):
+        return -1
```
Currently not supported and only used for progress logging

dimagimon commented on corehq/pillows/case.py:
```diff
@@ -6,15 +6,22 @@
 from pillowtop.listener import lock_manager
 from pillowtop.pillow.interface import ConstructedPillow
 from pillowtop.processors.elastic import ElasticProcessor
+from pillowtop.reindexer.change_providers.couch import CouchViewChangeProvider
```
F401 'CouchViewChangeProvider' imported but unused

dimagimon commented on corehq/pillows/case.py:
```diff
@@ -6,15 +6,22 @@
 from pillowtop.pillow.interface import ConstructedPillow
 from pillowtop.processors.elastic import ElasticProcessor
+from pillowtop.reindexer.change_providers.couch import CouchViewChangeProvider
+from pillowtop.reindexer.reindexer import get_default_reindexer_for_elastic_pillow, \
```
F401 'get_default_reindexer_for_elastic_pillow' imported but unused

dimagimon commented on corehq/pillows/case.py:
```diff
@@ -6,15 +6,22 @@
 from corehq.apps.change_feed.consumer.feed import KafkaChangeFeed
 from corehq.elastic import get_es_new
+from corehq.form_processor.backends.sql.dbaccessors import CaseReindexAccessor
+from corehq.pillows.mappings.case_mapping import CASE_MAPPING, CASE_INDEX, CASE_ES_TYPE, CASE_ES_ALIAS
```
F401 'CASE_ES_ALIAS' imported but unused
F401 'CASE_ES_TYPE' imported but unused
F401 'CASE_INDEX' imported but unused
F401 'CASE_MAPPING' imported but unused

snopoke committed:
fix couch case reindexer

snopoke committed:
[ci skip] rename class and docs

snopoke committed:
fix imports

snopoke committed:
Merge branch 'master' into sk/pagination

snopoke committed:
Merge branch 'master' into sk/pagination

snopoke committed:
use function to get doc_type_map

snopoke committed:
fix test mechanism to discard state

snopoke committed:
remove slug arg from super call

snopoke committed:
Merge branch 'master' into sk/pagination

millerdev commented on corehq/util/pagination.py:
```diff
@@ -55,14 +62,149 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
+    args = ListProperty()
+    kwargs = DictProperty()
+    retry = DictProperty()
+    progress = DictProperty()
```
Previously the `progress_info` property could store any type of object (as long as it was JSON-serializable). In practice the value was only ever a dict, but the design was intentionally open. Does this new definition constrain it to dict only?

snopoke replied:
Yes it does but I can change it to a `DefaultProperty` which is unconstrained. I actually prefer having it be a dict though since then you always know what you're going to get back. It also gives you more information about what the data in the progress is since each element would need a key which also serves as a descriptor., You can still store anything as a value so it doesn't restrict the options.

millerdev replied:
Yeah, thinking about it more, it makes sense to always use a dict.

millerdev commented on corehq/util/pagination.py:
```diff
@@ -55,14 +62,149 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
+    :param data_function: function to iterate over. Must return an list of data elements.
+    :param args_function: given the last data element of the previous ``data_function`` call and the previous
+    ``data_function`` args and kwargs return the next set of arguments for ``data_function``.
+    Initially called with ``None`` to get the fist set of arguments.
```
s/fist/first/

snopoke commented:
:white_check_mark:

snopoke commented on corehq/util/doc_processor/couch.py:
```diff
@@ -0,0 +1,135 @@
+    def data_function(**view_kwargs):
+        return db.view('all_docs/by_doc_type', **view_kwargs)
+
+    doc_types = sorted(doc_types)
```
@millerdev why did you sort these? It doesn't seem necessary for anything except perhaps testing and it means that you can't give priority to any doc types - which would be good for the reindexers.

millerdev replied:
They're sorted to create a stable key value (I believe originally `doc_types` was a dict). I think it would be fine to use the natural ordering of whatever data structure is passed in here and only sort for the key name (I'm making another note elsewhere about moving these particular iteration naming details into `BlobDbBackendMigrator`).

millerdev replied:
If you decide to change it to use the natural ordering of whatever `doc_types` sequence is passed in to allow prioritization, it should be documented that the sequence order must never change while the iteration is in progress since that could cause some doc types to be omitted completely if the iteration was stopped and then resumed with a different doc type ordering. Might want to also check that they didn't pass a dict or set since the iteration order of those types can change for [different interpreter runs](https://lwn.net/Articles/491788/).

millerdev commented:
Starting reviewing last night, but didn't get very far. Will continue where I left off today.

snopoke commented:
Thanks

millerdev commented on corehq/util/pagination.py:
```diff
@@ -55,14 +62,149 @@ def paginate_function(get_data_fn, next_args_fn, event_handler=PaginationEventHa
+        self.data_function = data_function
+        self.args_function = args_function
+        self.event_handler = event_handler
+        self.iteration_name = '{}/{}'.format(iteration_key, data_function.__name__)
```
What do you think of moving this out to `BlobDbBackendMigrator` and renaming `iteration_key` to `iteration_name` with a note in the docstring that `iteration_name` should be unique because it will be used to generate a unique identifier for persistent iteration state?

The reason I'm suggesting this is because the current way `ResumableDocsByTypeIterator` overwrites `data_function.__name__` seems like a bit of a hack, and obfuscates the real function's name, which could be useful in debugging, etc.

snopoke replied:
Yea, happy to do that.

snopoke replied:
probably worth leaving to a fllowup PR

millerdev commented on corehq/util/pagination.py:
```diff
@@ -0,0 +1,272 @@
+    by the ``args_provider``
+
+    :param data_function: function to paginate. Must return an list of data elements.
+    :param args_provider: An instance of the ``ArgsProvider`` class.
```
Might want to say something about what this is used for (provides args for `data_function`)

snopoke replied:
0888474

millerdev commented on corehq/util/pagination.py:
```diff
@@ -182,6 +195,31 @@ def _get_event_handler(self):
+        iteration has completed, and every retry document will be
+        yielded each time the iterator is stopped and resumed during the
+        retry phase. This method is relatively inefficient because it
+        forces the iteration state to be saved to couch. If you find
```
s/saved to couch/saved/

millerdev replied:
Hmm, I see this always saves state in the "meta" couchdb, so this is still correct. I haven't thought through this completely, but the original design of saving to the db in which docs are being iterated was a feature. It had been possible perform the same iteration on multiple different databases and their state would be indepdendent. Now it seems like the iteration state would be shared unless we somehow incorporate the database name/backend into `iteration_name`.

snopoke replied:
What's the use case for running the same iteration over different DB's?

millerdev replied:
No specific use case in mind right now, but multiple databases can exist having the same doc types. It's easier to reason about data consistency if the docs and the iteration state live in the same db.

snopoke replied:
And if the docs live in SQL?

millerdev replied:
I'd probably create table in the SQL database to store migration data.

benrudolph replied:
fwiw, and not sure if i'm speaking out of place here since i haven't been following the whole thing, but before i knew simon was working on this i was implementing the same thing with a similar approach. essentially making a ReusableResumableIterator where i just copied a bunch of the ResumableIterator code haha. anyways, one of my first changes was to move the state of the iterator to the meta database. as daniel mentioned though, i added the database name to the iterator_name to ensure that it would be unique in the meta db. while i thought keeping the iteration state in one time migrations made sense these were my reasons for saving the state in a separate db:

1. the migration state was now going to be run multiple times for reindexes (instead of just a one time thing) and something we'd probably want to have on a dashboard somewhere. creating a view for one database that got all the in-progress iterators over having a view in every database to view the status of the iterator seemed better

2. as already mentioned above, i didn't want to create a separate abstraction for iterating over SQL vs. Couch. ideally the reindexer would describe how to iterate and the iterator would iterate over it and save the state of iterator and not have to know about where that data lived

snopoke replied:
Thanks for your perspective Ben. I hadn't thought about a dashboard but
that would be nice.

I certainly don't plan on changing this behaviour. If the same iteration
needs to be run on multiple databases then the caller can ensure that the
iteration key is unique for the different DBs.

millerdev replied:
fwiw, changing it to save to "meta" has orphaned the status records for previous blob migrations. I doubt you care about that though.

snopoke replied:
What is that state needed for? I could write a migration to move them to
meta.

millerdev replied:
if the migration is run again it tells you it's already been run. it's not important enough to make a migration for it

millerdev commented on corehq/util/couch_doc_processor.py:
```diff
@@ -1,16 +1,39 @@
+class ResumableDocsByTypeArgsProvider(PaginatedViewArgsProvider):
+    def __init__(self, initial_view_kwargs, doc_types):
+        super(ResumableDocsByTypeArgsProvider, self).__init__(initial_view_kwargs)
+        self.doc_types = doc_types
```
Should probably be `self.doc_types = list(doc_types)` (make a copy) since it gets mutated.

snopoke replied:
7d2f383

millerdev commented on corehq/util/couch_doc_processor.py:
```diff
@@ -1,16 +1,39 @@
+        return last_args, last_view_kwargs
+
+
+def ResumableDocsByTypeIterator(db, doc_types, iteration_key, chunk_size=100, view_event_handler=None):
```
What do you think of renaming this to `resumable_docs_by_type_iterator` now that it's a function instead of a class? I was a little confused when reading the code until I realized that it had been changed to a function.

snopoke replied:
3893c34

millerdev commented on corehq/util/couch_doc_processor.py:
```diff
@@ -29,162 +52,34 @@ class ResumableDocsByTypeIterator(object):
+        'limit': chunk_size,
+        'include_docs': True,
+        'reduce': False,
+        'startkey': [doc_types[0]],
```
This will raise `IndexError` if the given `doc_types` is empty. I think that should not be an error condition, but instead create an iterator that yields zero items (pretty sure that's how it worked before).

snopoke replied:
Fair enough though passing in an empty list of doc types also seems like an error

snopoke replied:
Not going to implement the empty iterator, just going to document that and empty doc type list isn't supported.

snopoke replied:
f3dcc33

millerdev commented on corehq/util/pagination.py:
```diff
@@ -0,0 +1,272 @@
+    indefinitely unless it is removed with `discard_state()`.
+    :param data_function: function to iterate over. Must return an list of data elements.
+    :param args_provider: An instance of the ``ArgsProvider`` class.
+    :param item_getter: Function which can be used to get an item by ID. Used for retrying items that failed.
```
This complicates the constructor by adding a required argument that is only used by the optional retry feature. The generalization in this PR is making it less obvious how to use the retry feature: it requires this new `item_getter` and the doc ID must now be passed to `retry` instead of the object yielded from the iterator (doc).

It occurs to me that the retry feature could be implemented as a separate iterator/wrapper rather than conflating it with the resumable iterator. Usage would be something like this:
```python
def save_retries(data):
    # NOTE any other use of `resumable.progress_info` needs to
    # be aware of this so it doesn't stomp on the retry state.
    # To make that less error prone, could change progress_info
    # to be functions:
    #   resumable.get_progress_info(key)
    #   resumable.set_progress_info(key, value)
    progress_info = resumable.progress_info or {}
    progress_info["retry_data"] = data
    resumable.progress_info = progress_info

def item_getter(doc_id):
    try:
        return db.get(doc_id)
    except ResourceNotFound:
        raise SkipRetry

resumable = ResumableFunctionIterator(key, data_function, args_provider)
retries = (resumable.progress_info or {}).get("retries")

# `retries` and `save_retries` are optional arguments that can be
# omitted if persistence is not needed
retry_iter = RetryIterator(resumable, item_getter, retries, save_retries)

for doc in retry_iter:
    try:
        process(doc)
    except StaleDocError:
        retry_iter.retry(doc["_id"])
```

snopoke replied:
yes, I agree that's preferable. I was going to do that later but for now my priority is getting the reindexers done since that is blocking some other work.

millerdev replied:
Fine to do that later.

millerdev commented on corehq/util/couch_doc_processor.py:
```diff
@@ -29,162 +52,34 @@ class ResumableDocsByTypeIterator(object):
+    def item_getter(doc_id):
         try:
-            self.db.delete_doc(self.iteration_id)
+            return {'doc': db.get(doc_id)}
```
Not understanding this, and I can't find the place where this gets unwrapped. I thought the iterator was always dealing with doc dicts, not `{"doc": the_real_doc}`. I assume you're using the wrapper as a signal in case `db.get(doc_id)` returns `None`? Seems like it would be better to this:
```py
try:
    return db.get(doc_id)
except ResourceNotFound:
    raise SkipRetry
```

snopoke replied:
Actually the iterator iterates over couch view rows which get 'unwrapped' 4 lines above this. Returning None indicates that this doc should be skipped (same as before)

millerdev commented on corehq/util/tests/test_couch_doc_processor.py:
```diff
@@ -75,14 +79,14 @@ def test_resume_iteration_after_complete_iteration(self):
     def test_iteration_with_retry(self):
         itr = iter(self.itr)
         doc = next(itr)
-        self.itr.retry(doc)
```
I'm lamenting the loss of simplicity and aesthetics here 😞

snopoke replied:
Not quite sure what you mean by this and I can't see the full diff.

millerdev replied:
```diff
-        self.itr.retry(doc)
+        self.itr.retry(doc['_id'])
```
https://github.com/dimagi/commcare-hq/pull/12326/files#diff-a33b46af00c46590886b95c876a98ac2R93

millerdev commented on corehq/util/couch_doc_processor.py:
```diff
@@ -147,6 +147,42 @@ def should_process(self, doc):
 
+class ProcessorProgressLogger(object):
+    def progress_starting(self, total, previously_visited):
+        print("Processing {} documents{}: ...".format(
```
I see you removed the list of doc types being migrated here. That was a pretty useful sanity check output to verify that the selected migration was actually migrating the expected documents. Is that being logged somewhere else? If not, I'd suggest adding an optional `message=""` parameter here (default is empty) so it can be re-added.

millerdev replied:
bleh... didn't notice the class below, although I still may prefer the default message parameter to a whole extra subclass.

snopoke replied:
Since the generic iterator doesn't know about doc types anymore it's a bit hard to pass then to the logger.

snopoke committed:
[ci skip] more detailed docstring

snopoke committed:
copy doc_types list

snopoke committed:
rename function

snopoke committed:
add not about not supporting empty doc type list

snopoke commented:
followed up on a few of the smaller comments - can do a follow up PR for some of the other + some other stuff

millerdev commented:
I'm still in the middle of the review. This one is going to take a while (it's a big PR).

millerdev commented on corehq/util/doc_processor/couch.py:
```diff
@@ -0,0 +1,135 @@
+    All documents must live in the same couch database.
+
+    :param iteration_key: unique key to identify the document iterator
+    :param doc_types: Dict of `doc_type_name: model_class` pairs.
```
Should say something like "An ordered sequence. Each item in the sequence should be either a doc type class or a tuple `(doc_type_name_string, doc_type_class)`"

millerdev commented on corehq/util/doc_processor/sql.py:
```diff
@@ -0,0 +1,90 @@
+            return [next_db, datetime.min, None], {}
+
+
+def ResumableSqlModelIterator(iteration_key, reindex_accessor, chunk_size=100, event_handler=None):
```
Does it make sense to rename this like 3893c34d7c50fd31cf087121ff03998799b1780e?

millerdev commented on corehq/util/doc_processor/sql.py:
```diff
@@ -0,0 +1,90 @@
+        else:
+            from corehq.sql_db.config import PartitionConfig
+            partition_config = PartitionConfig()
+            self.db_list = partition_config.get_form_processing_dbs()
```
Is there any chance that the items in this list could ever change when an iteration is stopped? I assume that would invalidate the iteration state. Is that a concern?

snopoke replied:
not really - it would require a settings file change and the only thing that could change without a large migration is the ordering.

millerdev commented:
Kicked 2 build nodes (setup had failed)

millerdev commented:
Outstanding comments can be addressed in follow-up PR

millerdev commented on corehq/util/doc_processor/interface.py:
```diff
@@ -0,0 +1,314 @@
+        print("Processed {}/{} of {} documents in {} ({} remaining)"
+              .format(processed, visited, total, time_elapsed, time_remaining))
+
+    def progress_complete(self, total, processed, visited, previously_visited, filtered):
```
args order does not match [the call](https://github.com/dimagi/commcare-hq/pull/12326/files#diff-da7fddbada9ec9a3b6760d0f91093b4eR236). I'm seeing output like:
```
Processed 7055/3088965 of 600 documents in 0:00:45.943778 (5:34:29.397920 remaining)
```
